### PR TITLE
Support collection in MOVE and MOVEXY

### PR DESCRIPTION
**Pull request by UltimateGeek**
Provide support for doing a MOVE and MOVEXY on collections:
* If `from` is an invalid ID, log warning
* else if `from` is valid, current behavior
* else if `collection` is invalid, log a warning
* else move items in collection to `to` or `x`.`y`

This will enable creating buttons for a holder of generic widgets to both distribute its items out to the board as well as return them to their individual holders.

Wiki updates:
* [ ] Button MOVE & MOVEXY: add `collection` and indicate `from` get precedence.

Tutorial updates:
* [ ] https://github.com/ArnoldSmith86/virtualtabletop-library/pull/7

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
You can play around with it here: http://212.47.248.129:8157/ukl9 (or any other room on that server)

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3009/pr-test (or any other room on that server)

**Comment by 96LawDawg** [comment]
A few comments about this new capability.  When a generic widget is MOVEd to a holder it does (and should) gain the holder as a parent.  However, it keeps that parent even after being MOVEXY'd to another location.  So if I MOVE it, then MOVEXY it, then try to MOVE it back to the same parent, that does not work because the program thinks it is already in the parent.  Probably should remove the parent attribute after MOVEXY.

Also, the faces attribute of MOVE and MOVEXY do not seem to work on these generic widgets.  Simple demo room in http://212.47.248.129:3009/pr-test.  Also, I noticed that Ghetto did not load for the test server room.

**Comment by UltimateGeek** [comment]
> A few comments about this new capability. When a generic widget is MOVEd to a holder it does (and should) gain the holder as a parent. However, it keeps that parent even after being MOVEXY'd to another location. So if I MOVE it, then MOVEXY it, then try to MOVE it back to the same parent, that does not work because the program thinks it is already in the parent. Probably should remove the parent attribute after MOVEXY.
> 
> Also, the faces attribute of MOVE and MOVEXY do not seem to work on these generic widgets. Simple demo room in http://212.47.248.129:3009/pr-test. Also, I noticed that Ghetto did not load for the test server room.

Thank you for the feedback @96LawDawg :-)

File: client/js/widgets/button.js [hunk 1]
```diff
@@ -141,7 +141,11 @@ class Button extends Widget {
         const count = a.count || 999999;
 
         if(isValidID(a.from) && isValidID(a.to)) {
-          if(a.face === null && typeof a.from == 'string' && typeof a.to == 'string' && !widgets.get(a.to).children().length && widgets.get(a.from).children().length <= count) {
+          const item = widgets.get(a.from)
+          const dest = widgets.get(a.to)
+          if (item && item.p('type') == null && dest.p('type') == 'holder') {
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:146]: At least for `dest` this should error if `a.to` is an array.

File: client/js/widgets/button.js [hunk 2]
```diff
@@ -141,7 +141,11 @@ class Button extends Widget {
...
+            item.moveToHolder(dest);
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:147]: You only do the `moveToHolder`; not the `flip`, not the `bringToFront` or the `parent`. That's probably what LawDawg noticed.

Maybe find a better way to not duplicate the code? Easiest would probably to move the function inside the `forEach` to the top of the `MOVE(XY)` blocks and call that with a single widget?

File: client/js/widgets/button.js (hunk 1 above)
Comment (ArnoldSmith86) [reply client/js/widgets/button.js:146]: I also think this shouldn't be `== null` but rather `!= 'holder'`. Any reason behind this?

File: client/js/widgets/button.js [hunk 3]
```diff
@@ -163,16 +167,21 @@ class Button extends Widget {
       }
 
       if(a.func == 'MOVEXY') {
-        setDefaults(a, { count: 1, face: null, x: 0, y: 0 });
+        setDefaults(a, { count: 1, face: null, x: 0, y: 0});
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:170]: MY SPACE!!! :D

File: client/js/widgets/button.js [hunk 4]
```diff
@@ -163,16 +167,21 @@ class Button extends Widget {
...
-            if(a.face !== null && c.flip)
-              c.flip(a.face);
-            c.p('parent', null);
-            c.bringToFront();
-            c.setPosition(a.x, a.y, a.z || c.p('z'));
-            c.updatePiles();
-          }));
+          const item = widgets.get(a.from)
+          if (item && item.p('type') == null) {
+            item.setPosition(a.x, a.y, a.z || item.p('z'))
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:174]: Same stuff as above.

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
Sorry, I didn't review it because it felt like we were going a different route.

Larkob want to talk about buttons in the meeting tomorrow so maybe wait with changes until then.

**Comment by ArnoldSmith86** [comment]
We discussed a lot of Button changes in our last meeting and our proposal would be to add a property `collection` to `MOVE` and `MOVEXY` (see #60). That's not quite as simple as your proposed change, but would make it possible and would be more coherent across functions.

Is that good enough for you?

**Comment by UltimateGeek** [comment]
> We discussed a lot of Button changes in our last meeting and our proposal would be to add a property `collection` to `MOVE` and `MOVEXY` (see #60). That's not quite as simple as your proposed change, but would make it possible and would be more coherent across functions.
> 
> Is that good enough for you?

I'm not sure. I'd like to talk with through with you to see what you think, @ArnoldSmith86 :

I created a holder of little tokens for my game based on what I saw in the "Holders as Child" tutorial:

Screen Shot 2021-01-09 at 7 22 25 PM

The reasons I liked it:
* Each circle is a little holder to drop a generic widget on. The little circle holder changes color when hovering above it, so it's a very satisfying user experience for when a player moves a widget back to its holder.
* Each widget that can be dropped on each little holder is a single character for its visual (instead of an image). This minimizes the number of assets for the game.

So basically there's a one-to-one correspondence with the widgets to little holders.

What I was planning on doing for my board's cleanup button was to just add a MOVE for each individual widget to put it back into its holder.

But now I see a couple of other options:
1. replace the generic widgets with individual decks that each belong to each little holder. Each deck has a single card, with a single face that is just the single character.
   * Downside: more complicated holder-deck definition to create
2. Define a holder (with a deck) that is a single row with offset for holding the little widget cards.
   * Downside: widgets would slide out instead of being picked up individually from their spot when one is taken (doesn't mimic the IRL experience as well)

So, with all that said: if I keep my current design, could I use a _collection_ reference instead of the individual widget ID to MOVE? I could...but then I'd have to add SELECT for each of my MOVEs. The code would be simpler if I could just do a MOVE on the specific widget I want to move.

**Comment by UltimateGeek** [comment]
Here's a sample of the vtt code I wrote to take advantage of the changes in this PR right now. I can't think of a way to do this with collections that is just as simple and easy-to-maintain.

Holder:
```json
  "holder-cubes_1": {
    "type": "holder",
    "x": 5,
    "y": 5,
    "z": 11946,
    "id": "holder-cubes_1",
    "width": 15,
    "height": 15,
    "css": "border-radius: 30px;",
    "parent": "holder-cubes",
    "dropTarget": {
      "passenger": "yes"
    },
    "dropOffsetX": -5,
    "dropOffsetY": -5
  },
```
Widget:
```json
  "text-cube_1": {
    "text": "🎁",
    "css": "font-size:25px;color:black;display:flex;justify-content:center;align-items:center;text-align:center;",
    "width": 25,
    "height": 25,
    "x": -5,
    "y": -5,
    "z": 154,
    "id": "text-cube_1",
    "parent": "holder-cubes_1",
    "passenger": "yes"
  },
```
Cleanup Button:
```json
      {
        "func": "MOVE",
        "to": "holder-cubes_1",
        "from": "text-cube_1"
      },
```

**Comment by UltimateGeek** [comment]
If it's collections-only and not individual widgets, the cleanup button code would be this, right?, using DEFAULT collection:
```json
      {
        "func": SELECT",
        "property": "id",
        "text-cube_1"
      {
        "func": "MOVE",
        "to": "holder-cubes_1"
      },
      { "(repeat for the rest of the widgets)..."
```
I guess that's not too bad.

File: client/js/widgets/button.js (hunk 1 above)
Comment (UltimateGeek) [reply client/js/widgets/button.js]: fixed

File: client/js/widgets/button.js (hunk 2 above)
Comment (UltimateGeek) [reply client/js/widgets/button.js]: fixed

File: client/js/widgets/button.js (hunk 4 above)
Comment (UltimateGeek) [reply client/js/widgets/button.js]: fixed

**Comment by UltimateGeek** [comment]
Ready for review. The PR's http://212.47.248.129:3009/pr-test room has been updated with the new corresponding tutorial on MOVE / MOVEXY in ArnoldSmith86/virtualtabletop-library#7.

**Comment by larkob** [comment]
> {task} Widget Guide: Update the intro statement about Holders being "drop zones" for also generic widgets (not just cards)
Please remove this task. The "widget guide" is targeted towards people who look at their game components and wonder how to translate them to VTT widgets. That's why the page makes use or real-world examples and does not give too many details about VTT widgets. 

Perhaps the page title should be relabeled to make this point more clear. Anyhow, with regard to Holders, I believe they deserve their own wiki page as there's a ton of stuff to talk about.

**Comment by larkob** [comment]
Actually, I changed my mind. Let's keep the "Widget Guide", but make it clear that the current info is simply the "Introduction" part, rather than the details part, similar to how I structured the Buttons page. It then needs sections to explain each widget in detail, which will take me some time to update, so stay tuned.

**Comment by larkob** [comment]
> So, with all that said: if I keep my current design, could I use a collection reference instead of the individual widget ID to MOVE? I could...but then I'd have to add SELECT for each of my MOVEs. The code would be simpler if I could just do a MOVE on the specific widget I want to move.

Please see issue #61 which proposes substantial improvements to the SELECT function. Moving a specific widget would still require a prior SELECT, but that's true for pretty much every function.

We've also discussed an approach for performing functions with prior SELECT (or use of holders), which I've just documented as issue #72. Please have a look.

**Comment by larkob** [comment]
And I apologize for coming up with all those things at this late stage, but the general "button overhaul" discussion only happened quite recently.

**Comment by UltimateGeek** [comment]
> Actually, I changed my mind. Let's keep the "Widget Guide", but make it clear that the current info is simply the "Introduction" part, rather than the details part, similar to how I structured the Buttons page. It then needs sections to explain each widget in detail, which will take me some time to update, so stay tuned.

I updated the task's wording for modifying the wiki. Is that what you intended? If we're going to modify the Widget Guide's intro for this concern, it can be done right away. No need to wait on this PR.

**Comment by UltimateGeek** [comment]
> And I apologize for coming up with all those things at this late stage, but the general "button overhaul" discussion only happened quite recently.

No problem. I'm sorry I was unable to attend last week's meeting.

File: client/js/widgets/button.js [hunk 5]
```diff
@@ -143,48 +143,72 @@ export class Button extends Widget {
...
+        const debug = this.p('debug');
+        const _movexy = function(w) {
+          if(debug) console.log(`moving ${w.p('id')} to x:${a.x}, y:${a.y}`);
+          if(w.p('type') == 'card')
+            if(a.face !== null && w.flip)
+              w.flip(a.face);
+          w.p('parent', null);
+          w.bringToFront();
+          w.setPosition(a.x, a.y, a.z || w.p('z'));
+          if(w.p('type') == 'card')
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:199]: Same here. While right now only cards can pile, this should change in the future. And this should break moving piles onto cards/piles.

File: client/js/widgets/button.js [hunk 6]
```diff
@@ -143,48 +143,72 @@ export class Button extends Widget {
...
+            w.updatePiles();
+        };
+        setDefaults(a, { collection: 'DEFAULT', count: 1, face: null, x: 0, y: 0 });
+        if(a.from) {
+          if(!isValidID(a.from))
+            problems.push(`Warning: ${a.from} is not a valid widget ID`);
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:205]: Same.

File: client/js/widgets/button.js [hunk 7]
```diff
@@ -143,48 +143,72 @@ export class Button extends Widget {
...
+          else
+            this.w(a.from, source=>source.children().slice(0, a.count || 999999).reverse().forEach(c=> _movexy(c)));
+        } else if(! isValidCollection(a.collection)) {
+          problems.push(`Warning: ${a.collection} is not a valid collection`);
```
Comment (ArnoldSmith86) [inline client/js/widgets/button.js:209]: Same.

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
I don't mind merging this after a few changes. But since this doesn't only apply to `MOVE` and `MOVEXY`, I was thinking on a more general solution.

Something like extending the helper function `this.w` (which gets widgets) to take care of differentiating between `holder` and `collection`.

By the way, for `MOVE` this pretty much just becomes an alias of this (which is already possible in production):

```JSON
{
  "func": "SELECT",
  "property": "id",
  "value": "text-cube_1"
},
{
  "func": "SET",
  "value": "holder-cubes_1"
}
```

**Comment by larkob** [comment]
> > Actually, I changed my mind. Let's keep the "Widget Guide", but make it clear that the current info is simply the "Introduction" part, rather than the details part, similar to how I structured the Buttons page. It then needs sections to explain each widget in detail, which will take me some time to update, so stay tuned.
> 
> I updated the task's wording for modifying the wiki. Is that what you intended? If we're going to modify the Widget Guide's intro for this concern, it can be done right away. No need to wait on this PR.

What I meant is that I'd like the introduction to be targeted towards people that have no clue about how to properly use "widgets" in VTT, but want to learn how the things they are familiar with (boards, cards, dice, etc.) could be represented by widgets. So I start deliberately simple for a better learning curve. Then, in the (upcoming) details section about holders, we can talk about all the possibilities that holders can be used for.

**Comment by UltimateGeek** [comment]
> What I meant is that I'd like the introduction to be targeted towards people that have no clue about how to properly use "widgets" in VTT, but want to learn how the things they are familiar with (boards, cards, dice, etc.) could be represented by widgets. So I start deliberately simple for a better learning curve. Then, in the (upcoming) details section about holders, we can talk about all the possibilities that holders can be used for.

Ok. I'll just remove the task. I trust the content will be improved based on feedback, and will just focus more on the detailed reference material as it relates to my changes.

**Comment by larkob** [comment]
@UltimateGeek I'm not quite sure what the targeted scope of this PR is. Would you mind adding a summary to the [Upcoming Changes](https://github.com/ArnoldSmith86/virtualtabletop/wiki/Upcoming-Changes) page in the wiki?

**Comment by UltimateGeek** [comment]
> @UltimateGeek I'm not quite sure what the targeted scope of this PR is. Would you mind adding a summary to the [Upcoming Changes](https://github.com/ArnoldSmith86/virtualtabletop/wiki/Upcoming-Changes) page in the wiki?

Yes, will do. But I won’t have time until this weekend.

**Comment by RaphaelAlvez** [comment]
PR #373 should cover the issues addressed in this PR

**Pull request closed without merging**